tgui/packages/tgui/interfaces/NtosNotepad.tsx

```diff
@@ -1,39 +1,682 @@
+    />,
+  ];
+  const viewOptions = [
+    <MenuBarDropdown.MenuItemToggle
+      checked={showStatusBar}
+      {...getMenuItemProps('statusBar', 'Status Bar')}
+    />,
+  ];
+  const helpOptions = [
+    <MenuBarDropdown.MenuItem
+      {...getMenuItemProps('aboutNotepad', 'About Notepad')}
+    />,
+  ];
+  /* eslint-enable react/jsx-key */
+  const itemProps = {
+    openOnHover,
+    setOpenOnHover,
+    openMenuBar,
+    setOpenMenuBar,
+  };
+
+  return (
+    <Box className="NtosNotepad__MenuBar">
+      <MenuBarDropdown
+        entry="file"
+        openWidth="22rem"
+        display={<PartiallyUnderlined str="File" indexStart={0} />}
+        options={fileOptions}
+        {...itemProps}
+      />
+      <MenuBarDropdown
+        entry="edit"
+        openWidth="22rem"
+        display={<PartiallyUnderlined str="Edit" indexStart={0} />}
+        options={editOptions}
+        {...itemProps}
+      />
+      <MenuBarDropdown
+        entry="format"
+        openWidth="15rem"
+        display={<PartiallyUnderlined str="Format" indexStart={1} />}
+        options={formatOptions}
+        {...itemProps}
+      />
+      <MenuBarDropdown
+        entry="view"
+        openWidth="15rem"
+        display={<PartiallyUnderlined str="View" indexStart={0} />}
+        options={viewOptions}
+        {...itemProps}
+      />
+      <MenuBarDropdown
+        entry="help"
+        openWidth="17rem"
+        display={<PartiallyUnderlined str="Help" indexStart={0} />}
+        options={helpOptions}
+        {...itemProps}
+      />
+    </Box>
+  );
+};
+
+interface StatusBarProps {
+  statuses: Statuses;
+}
+
+const StatusBar = (props: StatusBarProps) => {
+  const { statuses } = props;
+  return (
+    <Box className="NtosNotepad__StatusBar">
+      <Box className="NtosNotepad__StatusBar__entry" minWidth="15rem">
+        Ln {statuses.line}, Col {statuses.column}
+      </Box>
+      <Box className="NtosNotepad__StatusBar__entry" minWidth="5rem">
+        100%
+      </Box>
+      <Box className="NtosNotepad__StatusBar__entry" minWidth="12rem">
+        NtOS (LF)
+      </Box>
+      <Box className="NtosNotepad__StatusBar__entry" minWidth="12rem">
+        UTF-8
+      </Box>
+    </Box>
+  );
+};
+
+const getStatusCounts = (text: string, selectionStart: number): Statuses => {
+  const lines = text.substr(0, selectionStart).split('\n');
+  return {
+    line: lines.length,
+    column: lines[lines.length - 1].length + 1,
+  };
+};
+
+type Statuses = {
+  line: number;
+  column: number;
+};
+
+const TEXTAREA_UPDATE_TRIGGERS = [
+  'click',
+  'input',
+  'paste',
+  'cut',
+  'mousemove',
+  'select',
+  'selectstart',
+  'keydown',
+];
+
+interface NotePadTextAreaProps {
+  maintainFocus: boolean;
+  text: string;
+  wordWrap: boolean;
+  setText: (string) => void;
+  setStatuses: (statuses: Statuses) => void;
+}
+
+class NotePadTextArea extends Component<
+  NotePadTextAreaProps,
+  { textAreaRef: RefObject<HTMLTextAreaElement> }
```

Comment:
This type is unnecessary, because your ref isn't in the state.

Reply:
👍